fix(elements): a couple of small fixes

gkalpak opened the pull request:
Several small fixes and refactorings. See individual commits for details.
This is an alternative to #31416 (with the feedback there addressed).

Jira issue: [FW-2006][1]

Fixes #30848.

##
NOTE:
The last commit is a fix to the public API. It should not affect users in any way (afaict), but we might still want to only merge that last commit to master-only. (The rest is good for both master and patch.)
Happy to move the last commit to a separate PR, if we decide it shouldn't go into patch.

_UPDATE: The last commit has been removed. See https://github.com/angular/angular/pull/36114#issuecomment-622378247._

[1]: https://angular-team.atlassian.net/browse/FW-2006

[FW-2006]: https://angular-team.atlassian.net/browse/FW-2006

mary-poppins commented:
You can preview 77ea12e at https://pr36114-77ea12e.ngbuilds.io/.

mary-poppins commented:
You can preview 04d17e1 at https://pr36114-04d17e1.ngbuilds.io/.

mary-poppins commented:
You can preview bb59493 at https://pr36114-bb59493.ngbuilds.io/.

antch commented:
@gkalpak As requested I created #36130 for the `firstChange` property issue. I also found another issue which can be seen in the repro -- the output events are not hooked to the DOM event listeners during the component lifecycle events `ngOnInit`, `ngOnChanges`, `ngAfterViewInit`, and `ngAfterContentInit`. Should I create a separate issue for that?

mary-poppins commented:
You can preview 2919a13 at https://pr36114-2919a13.ngbuilds.io/.

mary-poppins commented:
You can preview 5740be8 at https://pr36114-5740be8.ngbuilds.io/.

gkalpak commented:
Thx, @antch! Yes, please create a separate issue for output events not firing (since it is a different issue). You can use the same repro.

mary-poppins commented:
You can preview 2d059c3 at https://pr36114-2d059c3.ngbuilds.io/.

mary-poppins commented:
You can preview 88d6ab0 at https://pr36114-88d6ab0.ngbuilds.io/.

gkalpak commented on packages/elements/src/create-custom-element.ts:
```diff
@@ -136,31 +136,58 @@ export function createCustomElement<P>(
+
+        // Re-apply pre-existing values through the strategy.
+        preExistingValues.forEach(([propName, value]) => strategy.setInputValue(propName, value));
+      }
```
This new block of logic is what causes the payload size increase (~170B).

mary-poppins commented:
You can preview 68b1e4c at https://pr36114-68b1e4c.ngbuilds.io/.

mary-poppins commented:
You can preview 2295bc2 at https://pr36114-2295bc2.ngbuilds.io/.

andrewseguin reviewed (approved):
LGTM

IgorMinar commented on goldens/public-api/elements/elements.d.ts:
```diff
@@ -23,7 +23,7 @@ export declare interface NgElementStrategy {
     disconnect(): void;
     getInputValue(propName: string): any;
-    setInputValue(propName: string, value: string): void;
+    setInputValue(propName: string, value: any): void;
```
Is `any` absolutely necessary? I don't think it is, we could use `unknown` instead. right?

example: https://stackblitz.com/edit/setter-with-unknown-param

In general, we should be using `unknown` whenever possible instead of `any` (and we have been slowly introducing it into our code base and public api surface already).

Does that work for you?

gkalpak replied:
I changed to `any` to match the `getInputValue()` signature (which returns `any`).
Happy to go with `unknown` in both, although that has the slight potential of breaking people (in the unlikely case that they are implementing their own `NgElementStrategy`).

I added a fixup commit. PTAL

mary-poppins commented:
You can preview e82e1e3 at https://pr36114-e82e1e3.ngbuilds.io/.

mary-poppins commented:
You can preview 43b51aa at https://pr36114-43b51aa.ngbuilds.io/.

IgorMinar commented on goldens/public-api/elements/elements.d.ts:
```diff
@@ -22,8 +22,8 @@ export declare interface NgElementStrategy {
     disconnect(): void;
-    getInputValue(propName: string): any;
-    setInputValue(propName: string, value: string): void;
+    getInputValue(propName: string): unknown;
```
This is a breaking change with possibly a significant impact. Do we have some evidence that we could get away with this in google3 as well as externally?

We could intentionally make this change, as it's likely a small impact, and use it as a pilot for fixing other `any`s in our public API surface.

To a developer updating to v10, this will look like TypeScript compiler getting better at type checking, which might be a good justification for making this breaking change (in addition to the expected impact being minimal since this is not a frequently used API today compared to other APIs in Angular).

In any case, this should be documented as a breaking change.

IgorMinar reviewed (approved):
LGTM for API changes. I'm a bit worried about the change to the getter, but I'm ok with this being a guinea pig change which could be rolled back if the impact is too severe. Before we proceed lets verify that this doesn't break google3 and do a cursory research on github via code search. Thanks!

Reviewed-for: public-api

gkalpak commented:
At this point, I feel like the public API fix is not worth the time invested. Therefore, I am going to remove the public API changes and create an issue (in case we want to deal with it in the future).

---
For future reference:
I have search GitHub for custom implementation of `NgElementStrategy`. I have found at least three (that also implement `get/getInputValue()`):
- [spaceman27/angular9-d3: ng-zone-ng-element-strategy-factory.ts](https://github.com/spaceman27/angular9-d3/blob/c81b645a288f079be9326408e0e0be5ba2df3edb/libs/shared/utils/src/lib/elements-utils/ng-zone-ng-element-strategy-factory.ts)
- [tnicola/my-custom-element: elements-modified-zone-strategy.ts](https://github.com/tnicola/my-custom-element/blob/37ff1666579751a16589b9c92295401ca46ca73c/src/app/elements-modified-zone-strategy.ts)
- [jKelio/angular-element-example: custom-element-strategy.ts](https://github.com/jKelio/angular-element-example/blob/0cb8aba4ef0cea739120b09029f4224745223853/src/app/custom-strategy/custom-element-strategy.ts)

While I don't think these would break by the API changes, I can't be 100% sure. More importantly, the thing that matters more is where people might be using `get/setInputValue()`. This is quite hard to search for on GitHub (unless I miss something), since these terms are generic and yield too many results (unrelated to Angular Elements).

For reference, a earch for ["getInputValue("](https://github.com/search?q=%22getInputValue%28%22&type=Code) yields more than 300.000 results and for ["setInputValue("](https://github.com/search?q=%22setInputValue%28%22&type=Code) even more than that :grin:
Even trying to narrow them down by adding "angular" to the search, [I got 500+ results](https://github.com/search?q=%22setInputValue%28%22+angular&type=Code) and of course that does not guarantee that we are not missing valid results 🤷

mary-poppins commented:
You can preview 52d59af at https://pr36114-52d59af.ngbuilds.io/.

mary-poppins commented:
You can preview 7b7f205 at https://pr36114-7b7f205.ngbuilds.io/.

mary-poppins commented:
You can preview 0d94065 at https://pr36114-0d94065.ngbuilds.io/.

mary-poppins commented:
You can preview 3ab9f1d at https://pr36114-3ab9f1d.ngbuilds.io/.

mary-poppins commented:
You can preview aa39ff0 at https://pr36114-aa39ff0.ngbuilds.io/.

mary-poppins commented:
You can preview 1f32e73 at https://pr36114-1f32e73.ngbuilds.io/.

gkalpak commented on packages/elements/src/create-custom-element.ts:
```diff
@@ -182,20 +209,32 @@ export function createCustomElement<P>(
-    Object.defineProperty(NgElementImpl.prototype, property, {
-      get: function() {
-        return this.ngElementStrategy.getInputValue(property);
+  // TypeScript 3.9+ defines getters/setters as configurable but non-enumerable properties (in
+  // compliance with the spec). This breaks emulated inheritance in ES5 on environments that do not
+  // natively support `Object.setPrototypeOf()` (such as IE 9-10).
+  // Update the property descriptor of `NgElementImpl#ngElementStrategy` to make it enumerable.
+  Object.defineProperty(NgElementImpl.prototype, 'ngElementStrategy', {enumerable: true});
```
This is an example of the error that this addition fixes: https://circleci.com/gh/angular/angular/714093

gkalpak commented:
I have added two new commits at the end (currently f67b1b28126d9ae913083b4998b929114d73e6b3 and 1f32e73d0f094352c82764c9d002097dcad95153) to fix an issue with TS 3.9 and an issue with certain polyfills that do not call the constructor (and was affecting g3 - need new presubmit to confirm).

@andrewseguin, ptal.

AndrewKushnir commented:
[New Presubmit](http://test/OCL:312122416:BASE:312122454:1589827421784:758287a2).

AndrewKushnir commented on packages/elements/src/create-custom-element.ts:
```diff
@@ -136,31 +136,57 @@ export function createCustomElement<P>(
+              string, any
+            ] => [propName, (this as any)[propName]]);
+
+        // In some browsers (e.g. IE10), `Object.setPrototypeOf()` (which is required by some Custom
```
@gkalpak just want to check if IE 9/10 are affected only or this logic is needed for some other browsers as well? It looks like `Object.setPrototypeOf` is natively supported in modern browsers and given the fact that IE 9/10 will deprecated since v10, I'm wondering if this logic is needed?

gkalpak replied:
AFAICT, this isn;t needed for other browsers. However, this is going into v9 as well and also with IE<11 being deprecated doesn't mean we should break them :wink:
(Also, our current tests are failing without this fix; see https://github.com/angular/angular/pull/36114#discussion_r426700119.)

AndrewKushnir replied:
Sounds good, thanks @gkalpak 👍 
May be we can add a reference to [FW-2141](https://angular-team.atlassian.net/browse/FW-2141) (internal ticket to keep track of IE-specific code that we can refactor/cleanup later) as an inline comment?

gkalpak replied:
:+1: Added a [comment][1] on FW-2141.

[1]: https://angular-team.atlassian.net/browse/FW-2141?focusedCommentId=14257

AndrewKushnir replied:
Thanks @gkalpak!

AndrewKushnir commented:
FYI, Blueprint presubmit is successful. Started a [Global presubmit](https://test.corp.google.com/ui#id=OCL:312122416:BASE:312198700:1589853835175:fd069e3e) as well.

AndrewKushnir commented:
Global presubmit was successful as well 👍 (I've updated g3 status for this PR)

andrewseguin reviewed (approved):
LGTM

kara commented:
@gkalpak Can you rebase this? It's not applying cleanly to master for some reason. Also can you clarify if it should be in patch? Not sure what the resolution was for the last commit.

gkalpak committed:
fix(elements): correctly handle getting/setting properties before connecting the element

`createCustomElements()` creates some getters/setters for properties
corresponding to component inputs that delegate to the
`NgElementStrategy`. However, it is not guaranteed that the element's
`NgElementStrategy` will have been created when these getters/setters
are called, because some polyfills (e.g. `document-register-element`) do
not call the constructor.

Previously, trying to get/set input properties before connecting the
element to the DOM (via `connectedCallback()`) would fail due to
`NgElementStrategy` not being created.

This commit ensures that the `NgElementStrategy` is always created
before used inside the input property getters/setters (similar to how it
is done for other methods of `NgElement`).

Mentioned in https://github.com/angular/angular/pull/31416/files#r300326698.

gkalpak committed:
refactor(elements): simplify accessing `NgElementStrategy`

Previously, we had to check whether `NgElementStrategy` had been
instantiated before accessing it. This was tedious and error prone,
since it was easy to forget to add the check in new call sites.

This commit switches to using a getter, so that the check has to be
performed in one place and is transparent to call sites (including any
future ones).

gkalpak committed:
refactor(elements): remove unused imports and properties

- Remove imports that are not used.
- Remove private class properties that are not used outside of the
  constructor.

gkalpak committed:
test(elements): only declare helpers if needed

Previously, helper modules/components classes were declared even if the
tests were not run (because the environment did not support Custom
Elements for example).

This commit moves the declaration of the helpers inside the `describe()`
block, so they are not declared unnecessarily. This is in preparation of
adding more helpers that need to access variables declared inside the
`describe()` block.

gkalpak committed:
test(elements): clean up `TestStrategy` between tests

Previously, the `TestStrategy` `NgElementStrategy` used in
`createCustomElement()` tests was created once and re-used in each test
(due to complications related to how `customElements.register()` works).
As a result, the `TestStrategy` instance's state (e.g. inputs) could be
polluted from previous tests and affect subsequent ones.

This commit ensures the strategy instance is reset before each test.

gkalpak committed:
fix(elements): capture input properties set before upgrading the element

Previously, if an element started out as a regular `HTMLElement` (not a
Custom Element) and was later upgraded to a Custom Element, any
properties corresponding to component inputs that were set on the
element before upgrading it would not be captured correctly and thus not
reflected on the instantiated component.

This commit fixes it by ensuring that such properties are captured
correctly.

Jira issue: [FW-2006](https://angular-team.atlassian.net/browse/FW-2006)

Fixes #30848
Closes #31416

gkalpak committed:
fix(elements): do not break when the constructor of an Angular Element is not called

Previously, the correct behavior of Angular custom elements relied on
the constructor being called (and thus the `injector` property being
initialized). However, some polyfills (e.g. `document-register-element`)
do not call the constructor of custom elements, which resulted in the
`injector` property being undefined and the `NgElementStrategy` failing
to be instantiated.

This commit fixes it by being tolerant to the `injector` property being
undefined and falling back to the injector passed to the
`createCustomElement()` config.

NOTE:
We don't have proper tests exercising the situation where the
constructor is not called. For now this is tested using a Google
internal test suite (which is how this issue was caught).
This commit also adds a rudimentary unit test to emulate this situation.

gkalpak committed:
fixup! refactor(elements): simplify accessing `NgElementStrategy`

gkalpak commented:
I have removed the public API commit, so this is good for master _and_ patch. (I have updated the original comment.) I have also rebased and resolved conflicts.

mary-poppins commented:
You can preview dda5898 at https://pr36114-dda5898.ngbuilds.io/.
